kml/serdes_gpx.cpp

```diff
@@ -0,0 +1,273 @@
+#include "kml/serdes_gpx.hpp"
+
+#include "indexer/classificator.hpp"
+
+#include "coding/hex.hpp"
+#include "coding/point_coding.hpp"
+#include "coding/string_utf8_multilang.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "base/assert.hpp"
+#include "base/stl_helpers.hpp"
+#include "base/string_utils.hpp"
+#include "base/timer.hpp"
+
+#include <sstream>
+
+using namespace std::string_literals;
+
+namespace kml
+{
+namespace gpx
+{
+auto const kDefaultLang = StringUtf8Multilang::kDefaultCode;
+
+auto const kDefaultTrackWidth = 5.0;
+
+std::string PointToString(m2::PointD const & org)
+{
+  double const lon = mercator::XToLon(org.x);
+  double const lat = mercator::YToLat(org.y);
+
+  std::ostringstream ss;
+  ss.precision(8);
+
+  ss << lon << "," << lat;
+  return ss.str();
+}
+}
+
+GpxParser::GpxParser(FileData & data)
+: m_data(data)
+, m_categoryData(&m_data.m_categoryData)
+, m_attrCode(StringUtf8Multilang::kUnsupportedLanguageCode)
+{
+  ResetPoint();
+}
+
+void GpxParser::ResetPoint()
+{
+  m_name.clear();
+  m_description.clear();
+  m_org = {};
+  m_predefinedColor = PredefinedColor::None;
+  m_viewportScale = 0;
+  m_timestamp = {};
+  
+  m_color = 0;
+  m_styleId.clear();
+  m_mapStyleId.clear();
+  m_styleUrlKey.clear();
+  
+  m_featureTypes.clear();
+  m_customName.clear();
+  m_boundTracks.clear();
+  m_visible = true;
+  m_nearestToponym.clear();
+  m_nearestToponyms.clear();
+  m_properties.clear();
+  m_localId = 0;
+  m_trackLayers.clear();
+  m_trackWidth = gpx::kDefaultTrackWidth;
+  m_icon = BookmarkIcon::None;
+  
+  m_geometry.Clear();
+  m_geometryType = GEOMETRY_TYPE_UNKNOWN;
+}
+
+bool GpxParser::MakeValid()
+{
+  if (GEOMETRY_TYPE_POINT == m_geometryType)
+  {
+    if (mercator::ValidX(m_org.x) && mercator::ValidY(m_org.y))
+    {
+      // Set default name.
+      if (m_name.empty() && m_featureTypes.empty())
+        m_name[gpx::kDefaultLang] = gpx::PointToString(m_org);
+      
+      // Set default pin.
+      if (m_predefinedColor == PredefinedColor::None)
+        m_predefinedColor = PredefinedColor::Red;
+      
+      return true;
+    }
+    return false;
+  }
+  else if (GEOMETRY_TYPE_LINE == m_geometryType)
+  {
+    return m_geometry.IsValid();
+  }
+  
+  return false;
+}
+
+bool GpxParser::Push(std::string const & tag)
+{
+  m_tags.push_back(tag);
+  if (GetTagFromEnd(0) == "wpt")
+  {
+    m_geometryType = GEOMETRY_TYPE_POINT;
+  }
```

Comment:
```suggestion
    m_geometryType = GEOMETRY_TYPE_POINT;
```
We don't use {} for one-liners, it makes code easier to read.

Reply:
Fixed